SQL-561: Adding integration test harness

bucaojit opened the pull request:
Adding test harness for now, test case configuration files to come in tickets SQL-555, SQL-563.  In this PR as there are no test files, the MongoSQLIntegrationTest only checks for testcase files and returns.

Added sample test data in `resources/integration_test/testdata`, which is helpful for testing the loading feature with `./gradlew runDataLoader`

Query and DatabaseMetadata functions not fully tested as they require a change from ADL to implement `getSchemas()`.

bucaojit commented on src/integration-test/java/com/mongodb/jdbc/integration/testharness/IntegrationTestUtils.java:
```diff
@@ -0,0 +1,481 @@
+            List<Object> functionArgs, DatabaseMetaData databaseMetaData) throws SQLException {
+        int expectedColumnCount = 5;
+        if (functionArgs.size() != expectedColumnCount) {
+            throw new IllegalArgumentException(
```
Throwing IllegalArgumentException here and in other parts of the code when the configuration test files are incorrect.  Used instead of assert as I think assert is better used for verification errors of the results.

bucaojit replied:
From standup, team suggested looking into implementing dynamic tests for this:
https://www.baeldung.com/junit5-dynamic-tests

bucaojit replied:
Made changes to run tests as DynamicTest.  With the change, it no longer supports the BeforeEach/AfterEach methods which I was using for creating and closing the connection.  I also removed the pattern matching I was using as each test will be run as separate tests.  The test description will be used for the test name.

bucaojit commented on src/integration-test/java/com/mongodb/jdbc/integration/testharness/IntegrationTestUtils.java:
```diff
@@ -0,0 +1,481 @@
+
+        for (int i = 0; i < expectedRow.size(); i++) {
+            int columnType = rsMetadata.getColumnType(i + 1);
+            switch (columnType) {
```
Adding these types for now, may be enough for our integration test purposes.  Can add more if needed after writing and running integration test cases.

nbagnard commented on src/integration-test/java/com/mongodb/jdbc/integration/testharness/TestTypeInfo.java:
```diff
@@ -0,0 +1,54 @@
+    }
+
+    public static int nullableStringToInt(String type) {
+        switch (type) {
```
And here too, I think I would make this case insensitive.

bucaojit commented on src/integration-test/java/com/mongodb/jdbc/integration/MongoSQLIntegrationTest.java:
```diff
@@ -0,0 +1,36 @@
+import java.sql.Connection;
+import java.sql.SQLException;
+import java.util.List;
+import org.junit.Test;
```
Update made to use Junit 5 for integrationTest

bucaojit commented on src/integration-test/java/com/mongodb/jdbc/integration/MongoSQLIntegrationTest.java:
```diff
@@ -0,0 +1,36 @@
+    @Test
+    public void testRunQueryTests()
+            throws IOException, SQLException, IllegalAccessException, InvocationTargetException {
+        Connection conn = getBasicConnection(MONGOSQL);
```
Implemented the beforeeach / aftereach for db connection

bucaojit commented on src/integration-test/java/com/mongodb/jdbc/integration/MongoSQLIntegrationTest.java:
```diff
@@ -0,0 +1,36 @@
+            throws IOException, SQLException, IllegalAccessException, InvocationTargetException {
+        Connection conn = getBasicConnection(MONGOSQL);
+        String queryTestPrefix = "^Query.*";
+        List<Tests> tests = utils.loadTestConfigs(testDirectory);
```
Implemented the beforeall method for test config loading

bucaojit commented on src/integration-test/java/com/mongodb/jdbc/integration/testharness/DataLoader.java:
```diff
@@ -0,0 +1,63 @@
+
+        // Files may be large, read then load one file at a time
+        for (final File fileEntry : Objects.requireNonNull(folder.listFiles())) {
+            Yaml yaml = new Yaml(new Constructor(TestData.class));
```
Added a static yaml variable as mentioned in another comment, and used that for the yaml.load(is)

bucaojit commented on src/integration-test/java/com/mongodb/jdbc/integration/testharness/DataLoader.java:
```diff
@@ -0,0 +1,63 @@
+        for (final File fileEntry : Objects.requireNonNull(folder.listFiles())) {
+            Yaml yaml = new Yaml(new Constructor(TestData.class));
+            if (fileEntry.isFile()) {
+                InputStream is = new FileInputStream(fileEntry.getPath());
```
Thanks for that information.  I added a try with resources block.

bucaojit commented on src/integration-test/java/com/mongodb/jdbc/integration/testharness/IntegrationTestUtils.java:
```diff
@@ -0,0 +1,483 @@
+    public List<Tests> loadTestConfigs(String directory) throws FileNotFoundException {
+        List<Tests> tests = new ArrayList<>();
+        final File folder = new File(directory);
+        if (!folder.exists()) {
```
Made the change to return `tests`

bucaojit commented on src/integration-test/java/com/mongodb/jdbc/integration/testharness/IntegrationTestUtils.java:
```diff
@@ -0,0 +1,483 @@
+
+    private Tests processTestFile(String filename) throws FileNotFoundException {
+        Yaml yaml = new Yaml(new Constructor(Tests.class));
+        InputStream is = new FileInputStream(filename);
```
Thanks for that, added a try-with-resources

bucaojit commented on src/integration-test/java/com/mongodb/jdbc/integration/testharness/IntegrationTestUtils.java:
```diff
@@ -0,0 +1,483 @@
+    }
+
+    private Tests processTestFile(String filename) throws FileNotFoundException {
+        Yaml yaml = new Yaml(new Constructor(Tests.class));
```
Yes, that worked for this case and the dataloader.  Made those changes.

bucaojit commented on src/integration-test/java/com/mongodb/jdbc/integration/testharness/IntegrationTestUtils.java:
```diff
@@ -0,0 +1,483 @@
+        List<Object> metadataFunction = entry.meta_function;
+        assertTrue(metadataFunction != null && metadataFunction.size() > 0);
+        String functionName = (String) metadataFunction.remove(0);
+        switch (functionName) {
```
Sure, made the change to functionName and the cases

bucaojit commented on src/integration-test/java/com/mongodb/jdbc/integration/testharness/IntegrationTestUtils.java:
```diff
@@ -0,0 +1,483 @@
+            throws SQLException, InvocationTargetException, IllegalAccessException {
+        Method[] m = DatabaseMetaData.class.getMethods();
+        for (Method method : m) {
+            if (method.getName().equals(functionName)) {
```
Yes, made the change to use `equalsIgnoreCase`

bucaojit commented on src/integration-test/java/com/mongodb/jdbc/integration/testharness/TestTypeInfo.java:
```diff
@@ -0,0 +1,54 @@
+
+    public static int typesStringToInt(String type) throws IllegalAccessException {
+        for (Field field : Types.class.getFields()) {
+            if (field.getName().equals(type)) {
```
Sounds good, made the change

mattChiaravalloti commented on src/integration-test/java/com/mongodb/jdbc/integration/testharness/DataLoader.java:
```diff
@@ -0,0 +1,66 @@
+    public static void main(String[] args) throws IOException {
+        DataLoader loader = new DataLoader();
+        int rowsInserted = loader.loadTestData(TEST_DATA_DIRECTORY, MongoIntegrationTest.LOCAL_URL);
+        System.out.println("Inserted " + rowsInserted + " rows");
```
I wonder if we should log this per dataset instead of for the entire set of datasets. As in, `Inserted X rows into db.coll` for each dataset in the yaml files? Could be useful to see logged that each expected data set is restored? 

Maybe one day we could even extend the dataloader to only load data for datasets matching a pattern, so then this fine-tuned logging would become even more useful!

It may not be worth doing this per-dataset logging now, but it's something to think about.

bucaojit replied:
I added the logging per dataset, here's an example:
```
> Task :DataLoader.main()
Dropped integration_test.class
Dropped integration_test.grades
Inserted 4 rows into integration_test.class
Inserted 8 rows into integration_test.grades
```

mattChiaravalloti commented on src/integration-test/java/com/mongodb/jdbc/integration/testharness/models/TestEntry.java:
```diff
@@ -0,0 +1,31 @@
+    public String description;
+    public String sql;
+    public String skip_reason;
+    public Integer row_count;
+    public Boolean rowcount_gte;
```
```suggestion
    public Integer row_count;
    public Boolean row_count_gte;
```
[**super**-nit] Didn't catch this in the google doc, but should/could we make these two fields consistent by either including the `_` between `row` and `count` in both, or not including it in either? 😅

mattChiaravalloti commented on src/integration-test/java/com/mongodb/jdbc/integration/testharness/TestTypeInfo.java:
```diff
@@ -0,0 +1,54 @@
+    }
+
+    public static String typesIntToString(int type) throws IllegalAccessException {
+
```
```suggestion
```

mattChiaravalloti commented on src/integration-test/java/com/mongodb/jdbc/integration/testharness/TestTypeInfo.java:
```diff
@@ -0,0 +1,54 @@
+    }
+
+    public static int nullableStringToInt(String type) {
+
```
```suggestion
```

mattChiaravalloti commented on src/integration-test/java/com/mongodb/jdbc/integration/testharness/TestGenerator.java:
```diff
@@ -0,0 +1,136 @@
+import org.yaml.snakeyaml.Yaml;
+
+public class TestGenerator {
+    private static final String generatedTestDirectory = "resources/generated_test";
```
```suggestion
    private static final String GENERATED_TEST_DIR = "resources/generated_test";
```
same here re: idiomatic `static final` field names

mattChiaravalloti reviewed (changes requested):
Some thoughts for now. This is really great to see! I know there are still some things under construction and we may have dynamic test changes to come so I'll check back after more changes are pushed 👍

pmeredit commented on resources/integration_test/testdata/integration.yml:
```diff
@@ -0,0 +1,19 @@
+  - db: integration_test
+    collection: class
+    docs:
+      - {studentid: 10329,  name: "John", enrolled: true}
```
I think this is less of a worry because this is testing the driver and not mongosql. As long as it's some kind of number I think it's fine (i.e., no worries about smallest type).

bucaojit replied:
The numeric type for `studentid` that the `sqlGenerateSchema` command returns is `int`.
`testid` is also `int`.  `score` is `double`.

bucaojit commented on resources/integration_test/testdata/integration.yml:
```diff
@@ -0,0 +1,19 @@
+      - {studentid: 342,  name: "Jane", enrolled: false}
+      - {studentid: 303,  name: "Mike", enrolled: true}
+      - {studentid: 204323,  name: "Mary", enrolled: false}
+  - db: integration_test
+    collection: grades
+    docs:
+      - {studentid: 303, testid: 3, score: 84.5}
+      - {studentid: 10329, testid: 3,  score: 97.4}
+      - {studentid: 342, testid: 3,  score: 89.3}
+      - {studentid: 204323, testid: 3, score: 91.9}
+      - {studentid: 303, testid: 5, score: 87.5}
+      - {studentid: 10329, testid: 5,  score: 74.4}
+      - {studentid: 342, testid: 5,  score: 80.1}
+      - {studentid: 204323, testid: 5, score: 83.3}
```
Sure will do

bucaojit commented on src/integration-test/java/com/mongodb/jdbc/integration/MongoIntegrationTest.java:
```diff
@@ -0,0 +1,42 @@
+import java.sql.ResultSet;
+import java.sql.SQLException;
+
+public class MongoIntegrationTest {
```
Made the change to make MongoIntegrationTest abstract and moved the before/after test connection functions.

bucaojit commented on src/integration-test/java/com/mongodb/jdbc/integration/testharness/IntegrationTestUtils.java:
```diff
@@ -0,0 +1,483 @@
+        Pattern p = Pattern.compile(pattern);
+        for (Tests tests : testLists) {
+            for (TestEntry testEntry : tests.tests) {
+                if (testEntry.skip_reason != null || !p.matcher(testEntry.description).matches()) {
```
This was one of the suggestions to be able to run a test by pattern.  I think the idea was to use the pattern to run different sets of tests such as ExecuteQuery, PrepareExecute, PrepareOnly.  Or for example, if the binary data test takes a long time to run it can be separated and run in its own test function.
I'll change it to be case insensitive.

bucaojit commented on src/integration-test/java/com/mongodb/jdbc/integration/testharness/IntegrationTestUtils.java:
```diff
@@ -0,0 +1,483 @@
+    }
+
+    private ResultSet executeQuery(TestEntry entry, Connection conn) throws SQLException {
+        Statement stmt = conn.createStatement();
```
Added a try/finally to close the statement if not null

bucaojit commented on src/integration-test/java/com/mongodb/jdbc/integration/testharness/IntegrationTestUtils.java:
```diff
@@ -0,0 +1,483 @@
+            case "getUDTs":
+                return executeGetUDTs(metadataFunction, databaseMetaData);
+            default:
+                return executeDBMetadataHelper(functionName, metadataFunction, databaseMetaData);
```
Thanks for the suggestion.  
I made a change that handles the different parameter types besides String.  Handled in the same function instead of running different functions that were based on the parameter types.  I tested that the methods with non-string parameters are working.

mattChiaravalloti commented on src/integration-test/java/com/mongodb/jdbc/integration/testharness/TestGenerator.java:
```diff
@@ -0,0 +1,136 @@
+            expectedIsWritable.add(resultSetMetadata.isWritable(i));
+        }
+
+        testCase.put("description", description);
```
Should we also write the `sql` or `meta_function` that is the source of this test? And should we also include the `row_count`? And perhaps flags for whether or not the tests should be generated with `ordered` and/or `row_count_gte` set to true or false?

bucaojit replied:
Good catch, I added them to the generated output.

mattChiaravalloti commented on src/integration-test/java/com/mongodb/jdbc/integration/testharness/IntegrationTestUtils.java:
```diff
@@ -0,0 +1,428 @@
+                                                .toArray();
+                            }
+                        } else if (parameterType == String.class) {
+                            parameters[i] = (String) currentMetaInput;
```
Will this work when we want to pass null as an argument?

bucaojit replied:
I have an initial check if the `currentMetaInput` is null [here](https://github.com/mongodb/mongo-jdbc-driver/pull/62/files/1c69105a80c2a68a611e7414604c01b16fb7d866#diff-e017435f177b34f6c79fba493b695e25515551ac3ac287cc8554c7257bababbbR163) (line 163 if the diff doesn't load).  Is that what you were referring to?  If null it will set the parameter[i] to `null` and continue.

mattChiaravalloti replied:
oh perfect, sounds good 👍

mattChiaravalloti commented on resources/integration_test/testdata/integration.yml:
```diff
@@ -0,0 +1,20 @@
+      - {studentid: 10329,  name: "John", enrolled: true}
+      - {studentid: 342,  name: "Jane", enrolled: false}
+      - {studentid: 303,  name: "Mike", enrolled: true}
+      - {studentid: 204323,  name: "Mary", enrolled: false}
```
Let's manually include `_id` for every document like we do in the mongosql-rs spec test datasets. We could make them monotonically increasing ints like we do in that repo as well.

mattChiaravalloti commented on resources/integration_test/testdata/integration.yml:
```diff
@@ -0,0 +1,20 @@
+dataset:
```
Either as part of this ticket or a follow-up ticket (possibly SQL-563 which I have in progress), we should also expand the data restoration code to support creation of views. It should be pretty straightforward, including a `view` key to name a view where the `collection` is the source. Instead of `docs` we'll have `pipeline`. Let me know if you want to do that as part of this ticket or if you'd prefer to spin this work off to a follow-up ticket!

bucaojit replied:
That would be great to have.  Let's do this in a follow-up ticket.  Feel free to pick it up as part of SQL-563 if you would like.

mattChiaravalloti replied:
Ok cool. I'll plan on including that on 563 for now; I'll get on it after this is merged. When I get to it, I may still decided to bump it to a follow-up ticket, but we'll cross that bridge when we get there.

bucaojit commented on src/integration-test/java/com/mongodb/jdbc/integration/testharness/DataLoader.java:
```diff
@@ -0,0 +1,63 @@
+     * @throws FileNotFoundException
+     */
+    @SuppressWarnings("unchecked")
+    public int loadTestData(String dataDirectory, String url) throws FileNotFoundException {
```
Good idea!  I added a change to delete the collections before loading.

bucaojit commented on src/integration-test/java/com/mongodb/jdbc/integration/testharness/DataLoader.java:
```diff
@@ -0,0 +1,63 @@
+                }
+            }
+        }
+        return insertCounter;
```
I added a try catch block that deletes the collections on exception.

bucaojit commented on src/integration-test/java/com/mongodb/jdbc/integration/testharness/TestGenerator.java:
```diff
@@ -0,0 +1,136 @@
+public class TestGenerator {
+    private static final String generatedTestDirectory = "resources/generated_test";
+
+    public static void generateBaselineTestFiles(String description, ResultSet rs)
```
Using SnakeYaml to serialize the classes works as well.  When done that way, the fields are output in alphabetical order and I wanted to keep the ordering of description, sql/meta, expected_results, then the rest.  I can do more research on if there's a way to override the ordering, or maybe alphabetical is fine.

bucaojit replied:
I didn't have luck with overriding the alphabetical order that SnakeYaml was writing out.  The method I tried was modifying the [Representer](https://www.javadoc.io/doc/org.yaml/snakeyaml/1.29/org/yaml/snakeyaml/representer/Representer.html).  I do prefer the order it is currently vs alphabetical.  Let me know what you think.

nbagnard replied:
That's fine with me too :)

bucaojit commented on src/integration-test/java/com/mongodb/jdbc/integration/testharness/IntegrationTestUtils.java:
```diff
@@ -0,0 +1,483 @@
+                        return false;
+                    }
+                    break;
+                case Types.NULL:
```
Added the case for Types.OTHER.

mattChiaravalloti replied:
We'll probably want to file a follow-up ticket to support types that map to `Types.OTHER`

bucaojit replied:
Filed follow-up ticket to support types that map to `Types.OTHER`: [SQL-632](https://jira.mongodb.org/browse/SQL-632)

bucaojit commented on src/integration-test/java/com/mongodb/jdbc/integration/testharness/TestGenerator.java:
```diff
@@ -0,0 +1,136 @@
+        while (rs.next()) {
+            List<Object> row = new ArrayList<>();
+            for (int i = 1; i <= resultSetMetadata.getColumnCount(); i++) {
+                row.add(rs.getObject(i));
```
I tried serializing a BsonTimestamp and BsonArray, SnakeYaml was not able to serialize it well.  It printed out yaml tags without the values.  More work and investigation will need to be done to test the Types.OTHER.  As this PR is getting to be a bit much I think I will create a separate ticket to address this if that is alright with you.

nbagnard replied:
Sounds good, then you'll be able to check compareRows for OTHER and see if it work well too at this time. 👍 
Could you add TODO or another very clear comment the OTHER is not well supported for now?

bucaojit commented on src/integration-test/java/com/mongodb/jdbc/integration/MongoIntegrationTest.java:
```diff
@@ -0,0 +1,42 @@
+import java.sql.SQLException;
+
+public class MongoIntegrationTest {
+    public static final String LOCAL_URL = "mongodb://localhost:28017";
```
I added a note about needing the local mongodb and ADL.
I also added new environment variables needed to connect to that local instance.  I will need to add them as well to the Evergreen host image before actually adding test cases for mongosql.  I'll check with you on that later.

bucaojit commented on src/integration-test/java/com/mongodb/jdbc/integration/MongoSQLIntegrationTest.java:
```diff
@@ -0,0 +1,52 @@
+public class MongoSQLIntegrationTest extends MongoIntegrationTest {
+    public static final String MONGOSQL = "mongosql";
+    public static final String testDirectory = "resources/integration_test/tests";
+    private final IntegrationTestUtils utils = new IntegrationTestUtils();
```
Made the change to make them `static`

mattChiaravalloti commented on src/integration-test/java/com/mongodb/jdbc/integration/testharness/IntegrationTestUtils.java:
```diff
@@ -0,0 +1,429 @@
+                }
+                if (method.getParameterCount() == 0) {
+                    return (ResultSet) method.invoke(databaseMetaData);
+
+                } else {
```
```suggestion
                }
```
[nit] Since we `return` in the `if` block, let's drop the `else` here and move everything over one level of indentation.

mattChiaravalloti commented on src/integration-test/java/com/mongodb/jdbc/integration/testharness/IntegrationTestUtils.java:
```diff
@@ -0,0 +1,429 @@
+        }
+    }
+
+    private static void validateRowCount(
+            int expectedRowCount, TestEntry testEntry, Integer actualRowCounter, ResultSet rs)
+            throws SQLException {
+        if (actualRowCounter == null) {
+            actualRowCounter = MongoIntegrationTest.countRows(rs);
+        }
+        if (testEntry.row_count_gte != null && testEntry.row_count_gte) {
+            assertTrue(actualRowCounter >= expectedRowCount);
+        } else {
+            assertEquals(java.util.Optional.ofNullable(actualRowCounter), expectedRowCount);
```
```suggestion
    private static void validateRowCount(
            TestEntry testEntry, Integer actualRowCounter, ResultSet rs)
            throws SQLException {
        if (actualRowCounter == null) {
            actualRowCounter = MongoIntegrationTest.countRows(rs);
        }
        if (testEntry.row_count_gte != null && testEntry.row_count_gte) {
            assertTrue(actualRowCounter >= testEntry.row_count);
        } else {
            assertEquals(java.util.Optional.ofNullable(actualRowCounter), testEntry.row_count);
```
[nit] I don't think we should treat `testEntry.row_count` special by making it separate argument from `testEntry`, especially if we don't do the same for `testEntry.row_count_gte`. What I mean is, the signature should either be:
```
private static void validateRowCount(int expectedRowCount, Boolean expectedRowCountGte, Integer actualRowCount, ResultSet rs)
```
or
```
private static void validateRowCount(TestEntry testEntry, Integer actualRowCount, ResultSet rs)
```

I don't have a preference between the two, but think it is odd to pass `(testEntry.row_count, testEntry, ...)`.

bucaojit replied:
Ok sounds good.  I implemented the latter: 
```
private static void validateRowCount(TestEntry testEntry, Integer actualRowCount, ResultSet rs)
```

mattChiaravalloti commented on src/integration-test/java/com/mongodb/jdbc/integration/testharness/IntegrationTestUtils.java:
```diff
@@ -0,0 +1,429 @@
+            // Handle expected field being null
+            if (expectedRow.get(i) == null) {
+                if (actualRow.getObject(i + 1) == null) {
+                    break;
```
Why do we break here? Doesn't this mean there is a null at position `i` in the expected result but that's the only place it is expected? As in, `expected_result: [1, 2, null, 3]` why would we want to break on the third value? Or am I misunderstanding what's going on here?

bucaojit replied:
That's a good catch.  It's supposed to be `continue` not `break`.  I must have gotten mixed up with seeing the `break`s in the switch statement.

mattChiaravalloti commented on src/integration-test/java/com/mongodb/jdbc/integration/testharness/TestGenerator.java:
```diff
@@ -0,0 +1,145 @@
+            testCase.put("sql", testEntry.sql);
+        }
+        testCase.put("expected_result", result);
+        testCase.put("row_count", testEntry.row_count);
```
Should this value also be generated? That way it conveniently updates if we ever update our datasets?

mattChiaravalloti reviewed (changes requested):
Nice changes! 💯  I have a few more ideas of varying size but nothing too major 🤞

nbagnard commented on src/integration-test/java/com/mongodb/jdbc/integration/testharness/TestGenerator.java:
```diff
@@ -0,0 +1,145 @@
+    public static void main(String[] args)
+            throws SQLException, IOException, InvocationTargetException, IllegalAccessException {
+        MongoSQLIntegrationTest integrationTest = new MongoSQLIntegrationTest();
+        Connection conn = integrationTest.getBasicConnection();
```
Looks like you missed closing the connection here. Can you add a try with-resource here?

bucaojit commented on src/integration-test/java/com/mongodb/jdbc/integration/testharness/DataLoader.java:
```diff
@@ -0,0 +1,110 @@
+
+    public DataLoader(String dataDirectory, String url) throws IOException {
+        this.datasets = new ArrayList<>();
+        this.collections = new HashSet<>();
```
Thanks for that information.  I implemented the `equals()` and `hashCode()` methods and confirmed it passes the `testPairHash()` test.

bucaojit commented on src/integration-test/java/com/mongodb/jdbc/integration/MongoIntegrationTest.java:
```diff
@@ -0,0 +1,17 @@
+import java.sql.ResultSet;
+import java.sql.SQLException;
+
+public abstract class MongoIntegrationTest {
```
Right, before there was the connection and @before/after methods in the abstract class.  Which is no longer the case after the latest set of changes.
I have removed the abstract  `MongoIntegrationTest` class.  We can re-add it later if we add shared functionality.

bucaojit commented on src/integration-test/java/com/mongodb/jdbc/integration/MongoSQLIntegrationTest.java:
```diff
@@ -0,0 +1,61 @@
+@TestInstance(TestInstance.Lifecycle.PER_CLASS)
+public class MongoSQLIntegrationTest extends MongoIntegrationTest {
+    static final String URL = "jdbc:mongodb://localhost";
+    public static final String LOCAL_MDB_URL =
```
It's used in DataLoader.java.  I've moved variable over to DataLoader for clarity.

bucaojit commented on src/integration-test/java/com/mongodb/jdbc/integration/testharness/DataLoader.java:
```diff
@@ -0,0 +1,110 @@
+            }
+        } catch (MongoException ex) {
+            dropCollections();
+            ex.printStackTrace();
```
Yes that sounds good.  I made the change to re-throw the exception.

bucaojit commented on src/integration-test/java/com/mongodb/jdbc/integration/testharness/IntegrationTestUtils.java:
```diff
@@ -0,0 +1,429 @@
+                assertTrue(found);
+            }
+        }
+        if (testEntry.row_count != null) {
```
The `validateRowCount` method uses the value `actualRowCounter` which can be set during the row validation step.  
If not set, then the validateRowCount iterates through the rowset and counts the rows.  One way I can think of to move validateRowCount above the row validation is to have copy the results to a list as it iterates through the rows and use that list in the row validation.  I'm not sure if that adds too much value as it adds having to copy the results.  WDYT?

nbagnard replied:
oh, no, sorry, I forgot that we don't know the row count and have to go through the whole rs. 
There is one thing however that you might need to address : Technically, if I only want to validate the row count and not the rows themselves, it won't work right now because it will not go through the result set unless testEntry.expected_result != null, correct?

bucaojit replied:
No problem.
If `actualRowCounter` is still null then the `validateRowCount()` method will go ahead and count the rows of the resultset itself [here](https://github.com/mongodb/mongo-jdbc-driver/pull/62/files#diff-e017435f177b34f6c79fba493b695e25515551ac3ac287cc8554c7257bababbbR365).
Now that I'm looking at `actualRowCounter` I see I should move it's initial null check to before the `while (rs.next()) {` loop, I'll do that now.

bucaojit commented on src/integration-test/java/com/mongodb/jdbc/integration/testharness/IntegrationTestUtils.java:
```diff
@@ -0,0 +1,429 @@
+                    }
+                    break;
+                case Types.OTHER:
+                    if (expectedRow.get(i) != actualRow.getObject(i + 1)) {
```
I see.  I think I'll comment this case out for now and add the TODO to address it for SQL-632.

mattChiaravalloti reviewed (changes requested):
Looks great! The only bit of work that remains is plugging together this and SQL-566. We need an evergreen task that runs the script and then runs the integration tests. It should be pretty simple to add this task given the building blocks. The evergreen task feels linked to this work, but if you want to file a follow-up ticket that also makes sense. Your call.

bucaojit commented:
Made the change to run the local adl script and dataloader in the evergreen task.  Also added an empty run_adl.sh script which will be overridden with SQL-566.
Added a command in the dataloader to generate the schema for the database after it loads the data.

mattChiaravalloti reviewed (approved):
lgtm! 🔥 this is excellently done and I'm very excited to have it up and running going forward

nbagnard reviewed (approved):
LGTM! 👍
